xxh3.h

```diff
@@ -1509,18 +1506,47 @@ XXH_FORCE_INLINE void XXH3_initCustomSecret(xxh_u8* XXH_RESTRICT customSecret, x
      */
     XXH_ASSERT(kSecretPtr == kSecret);
 
-    for (i=0; i < nbRounds; i++) {
-        /*
-         * The asm hack causes Clang to assume that kSecretPtr aliases with
-         * customSecret, and on aarch64, this prevented LDP from merging two
-         * loads together for free. Putting the loads together before the stores
-         * properly generates LDP.
-         */
-        xxh_u64 lo = XXH_readLE64(kSecretPtr + 16*i)     + seed64;
-        xxh_u64 hi = XXH_readLE64(kSecretPtr + 16*i + 8) - seed64;
-        XXH_writeLE64(customSecret + 16*i,     lo);
-        XXH_writeLE64(customSecret + 16*i + 8, hi);
+#if (XXH_VECTOR == XXH_AVX512)
+    XXH_STATIC_ASSERT((XXH_SECRET_DEFAULT_SIZE & 63) == 0);
+    {   int const nbRounds = XXH_SECRET_DEFAULT_SIZE / sizeof(__m512i);
+        XXH_ALIGN(64) const __m512i* const src  = (const __m512i*) kSecret;
+        XXH_ALIGN(64)       __m512i* const dest = (      __m512i*) customSecret;
+                            __m512i  const seed = { (long)seed64, -(long)seed64, (long)seed64, -(long)seed64,
+                                                    (long)seed64, -(long)seed64, (long)seed64, -(long)seed64 };
+        for (i=0; i < nbRounds; ++i) dest[i] = _mm512_add_epi64(src[i], seed);
+    }
+#elif (XXH_VECTOR == XXH_AVX2)
+    XXH_STATIC_ASSERT((XXH_SECRET_DEFAULT_SIZE & 31) == 0);
+    {   int const nbRounds = XXH_SECRET_DEFAULT_SIZE / sizeof(__m256i);
+        XXH_ALIGN(64) const __m256i* const src  = (const __m256i*) kSecret;
+        XXH_ALIGN(64)       __m256i* const dest = (      __m256i*) customSecret;
+                            __m256i  const seed = { (long)seed64, -(long)seed64, (long)seed64, -(long)seed64 };
+        for (i=0; i < nbRounds; ++i) dest[i] = _mm256_add_epi64(src[i], seed);
+    }
+#elif (XXH_VECTOR == XXH_SSE2)
+    XXH_STATIC_ASSERT((XXH_SECRET_DEFAULT_SIZE & 15) == 0);
+    {   int const nbRounds = XXH_SECRET_DEFAULT_SIZE / sizeof(__m128i);
+        XXH_ALIGN(64) const __m128i* const src  = (const __m128i*) kSecret;
+        XXH_ALIGN(64)       __m128i* const dest = (      __m128i*) customSecret;
+                            __m128i  const seed = { (long)seed64, -(long)seed64 };
```

Comment:
Non-standard extension, will truncate seed64 on Windows or 32-bit

Reply:
Indeed, the `long` type is problematic, as its size, 32 or 64-bit, can be different depending on platform.
So it is best avoided.

Reply:
I want to go back in time and slap whoever thought up with making type sizes platform-dependent.

Java got it right.

C did not.

| Type | \*nix 64-bit | Windows 64-bit | 32-bit | 8-bit/16-bit |
|:----|---:|---:|---:|----:|
| `char` | 8 | 8 | 8 | 8 |
| `short` | 16 | 16 | 16 | 16 |
| `int` | 32 | 32 | 32 | 16 |
| `long` | 64 | 32 | 32 | 32 |
| `long long` | 64 | 64 | 64 | 64 |

Also add `char` being signed on x86 and unsigned on ARM/PPC/MIPS, and the standard type system is useless.

`stdint.h` or `limits.h` for C89 are the only reliable options.

Reply:
yes, c type system is so messy~ I add a `xxh_i64` for portable `int64_t`.